### ISSUE TITLE
Service scale-to-zero (fka "Socket activation")

### ISSUE THREAD

**Issue by KyleAMathews** [kubernetes/kubernetes#484]
I want to start this issue to drum up support for adding socket activation support to Kubernetes. My CTO at the last place I worked at, @davidstrauss, invented this technique.

Normally for a service to be available, its daemon has to be running and listening on a socket. With socket activation, another service, generally SystemD, listens for all incoming traffic and on an incoming request, passes the request to its service, waking it if necessary.

By idling inactive containers, there can be significant savings on memory. At Pantheon which hosts Drupal and Wordpress sites, at any time some ~80-90% of containers are idled waiting for traffic. This means they can proportionally increase the density of containers resulting in significant operations and server cost savings.

This is of course not a technique for everyone. You need to have services which won't get any use for significant periods of time and you can accept a few second delay from the initial waking time. But for many people, especially those running multi-tenant architectures, this would be a very valuable feature if baked in and easy to use and help Kubernetes be even more elastic.

Some more information on socket activation:
- http://0pointer.de/blog/projects/socket-activated-containers.html
- http://www.slideshare.net/warpforge/php-at-density-and-scale

**Comment by bgrant0607** [comment]
There was a proposal: #3247.

**Comment by bgrant0607** [comment]
Another proposal: https://github.com/kubernetes/kubernetes/pull/29471

**Comment by davidopp** [comment]
ref/ #42255

**Comment by warmchang** [comment]
ref: https://github.com/openshift/service-idler

**Comment by aledbf** [comment]
ref: https://github.com/deislabs/osiris/

**Comment by Zetanova** [comment]
any progress?

would really be great to socket activate pods to use them for webhooks/serverless services.

**Comment by wreed4** [comment]
Any solutions to this yet?  I'm also looking for a graceful solution to this.  I know GoDaddy built something like this as well which I thought they were going to open-source, but it doesn't seem like they have.

**Comment by Zetanova** [comment]
@remram44 socket activation is a specific technic, used in systemd.
https://0pointer.de/blog/projects/socket-activation.html

The host process creates/opens the socket (tcp port) and listen to traffic.
On the first connection, the host process starts the real-process and hands over the already open socket files 
with environment vars.

The k8s pod-pause container could do the same as systemd.

**Comment by Zetanova** [comment]
It is not about workload only, there can be sidecars for management like sshd, rsync and others in the pod
and custom k8s admission controllers could be started like it too.

@remram44 Yes, you are right that for serverless functions there are already good solutions out, like [knative](https://knative.dev/)

**Comment by remram44** [comment]
I feel like you are talking about something completely different from this issue, and something that doesn't need any support in Kubernetes at all. You can do socket activation of your software if you want to, container or not.

**Comment by Zetanova** [comment]
The issue is:
There is zero possibility to use the original container image in a pod,
suppress the health-check and only start the container on first tcp connection.

Places where socket activation could be in k8s used:
- k8s admission controllers
- rarely used sidecars like for backup, management access, debug and metrics.
- obsolete services where all consumers moved to the new service endpoint
- maybe others 

docker itself has some support for it.
systemd starts waits for the tcp connection and starts the container. 

The idea is that k8s could offer the same socket activation pattern

**Comment by manelclos** [comment]
How is this different from the (alpha) `HPAScaleToZero` feature gate? https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

More info in this issue: https://github.com/kubernetes/kubernetes/issues/69687

**Comment by remram44** [comment]
You could use that with a metric from your Ingress controller, to scale up from 0 when requests come in. It might be possible to do that today with the existing Ingress controllers. Of course those first few requests would *fail* until the pods are available.

**Comment by sftim** [comment]
Would anyone be willing to turn this into an [enhancement proposal](https://www.k8s.dev/resources/keps/)?

**Comment by thockin** [comment]
I updated the title on this.  I am not against considering it, but it's not as easy as it might sound, I think, especially if you consider things like external traffic policy

**Comment by ngrigoriev** [comment]
If it was possible to somehow transfer a TCP connection from one process (in one container) to another, then potentially an init container would be able to do the trick - shut down all pods, let one replica restart and let the init container to wait for the connection(s) before starting the main container of the first replica. But I doubt it is feasible.

I can imagine that something like this can be achieved by a smart filter/router on the API gateway. Basically, allow upstream to scale down to 0 replicas, accept the request and then scale up to at least one replica and wait until it is ready.

**Comment by thockin** [comment]
Sitting in an init container offers no practical advantage over leaving one
replica.  The resources (CPU, ram) are still consumed.

To scale to zero SOMEONE has to listen. An ingress proxy works well.
Service VIPs are harder.

**Comment by Zetanova** [comment]
@ngrigoriev  systemd made it view years ago, the host process opens the socket and on first connections starts the process with well known ENV variables set with the socket file path

Here are view links about the topic:
https://docs.docker.com/engine/reference/commandline/dockerd/
https://mgdm.net/weblog/systemd-socket-activation/
https://blog.developer.atlassian.com/docker-systemd-socket-activation/

The problem in k8s would be that it would need to support it natively.
Its simply useless and expensive to make it in-container

The `k8s.gcr.io/pause` image could implement it by create a socket on a mount 
and emits an API event when the first connection. 
k8s controller could then start the pod with the well-known `LISTEN_` ENV variables and the mounted socket files.

It would be in theory possible to make it over an init-container by simple holding the startup until the first connection
and extend the health check timeout. But the problem then would be to pass well-known `LISTEN_` ENV variables to the container.

**Comment by thockin** [comment]
Systemd works because it knows the connection was established to this
machine and this port. Handing off a file descriptor is easy in that
context. The 5-tuple is unchanged.

K8s doesn't know which node to run the new pod on until the traffic has
arrived. You can't be guaranteed to start a pod on that specific node. If
you start a pod elsewhere, the routing layer needs to know about it.

So you need a different mechanism. For example, the sending-side needs to
pause the traffic upon TX, wait for a pod, then route to it.  Now you have
to get that signal out from the node(s) to the upscaler.

Not impossible but not as easy has passing an FD.

**Comment by Zetanova** [comment]
@thockin there is a k8s.gcr.io/pause image that is always started for the pod
the socket activation would only be for the containers in the pod and not the pod environment itself.
the pod itself would still be created and k8s.gcr.io/pause would still be started

It makes a difference to run a legacy service version or API Admission Controller  basically only with <2MB memory footprint

**Comment by thockin** [comment]
First, pause is a runtime implementation detail - don't rely on it.

In order to schedule a pod, you need to set requests, at that point you
might as well leave 1 pod running - you have already incurred the cost.

**Comment by sftim** [comment]
This enhancement really could be useful. One benefit to a cluster operator is better bin packing. Provided that the background work is scheduled (Linux scheduler) and prioritised appropriately, the hit to latency maybe worth the savings in being able to run fewer nodes overall.

You could even use this to run an aggregated API. If there are no open watches, does the aggregated API need to be running all the time? Perhaps not.

**Comment by sftim** [comment]
_If_ we get a slew of somewhat related features then I could see this working another way.

Imagine a set of endpoints for a workload where a Pod for the app is “running”, but actually suspended. A connection arrives and this triggers a launch for the app container. Adding a new container, especially one that requests memory and CPU, is an in-place resize that may then trigger eviction of lower-priority Pods.

Perhaps the kubelet normally treats a socket-activated container as using its request. However, for a certain Pod priority threshold - I'd pick `0` as priorities can be negative - the kubelet overcommits the memory resource and allows both the socket-activated Pod and an evictable batch Job to compete for the same memory. The moment that the socket activation happens, the activator notifies the kubelet to act. The low-priority batch job now either risks an OOM kill, or plain gets evicted with no change to discuss it first.



---

We might specify that the socket activation must meet some requirements, either to simplify initial implementation or long term.Some examples:
- only one listening port
- TCP or UDP only
- the socket-activated container must have a memory limit equal to its request (to simplify eviction decisions)
- restrictions on what volumes may be mounted, at least initially
- only allow specifying container images by a content hash
- the socket-activated container is not allowed to use ServiceAccount tokens (there's a lead time on obtaining these)

**Comment by sftim** [comment]
This would need an enhancement proposal. I propose we set a deadline (at least weeks away).

If by that deadline we have a KEP issue logged and a PR in draft to add an initial KEP document, then we can link from the KEP issue here, and close this.
If by that deadline we don't have those things, we should close it as not implemented. Someone who wants to revisit the idea would be welcome to make a formal enhancement proposal at any time.

**Comment by sathieu** [comment]
This feature would be useful.

**Comment by jeacott1** [comment]
if its http(s) request activation then you can either build your own or tweak nginx to hold new connections while it triggers a new target container and then forwards once available. A map of existing connections can be kept in something like redis. This can also work well as a simple dynamic discovery mechanism by directly looking up srv records based on headers, in lieu of more elaborate setups like service mesh.

**Comment by bhack** [comment]
Keep alive!

**Comment by ShadowJonathan** [comment]
Why is this stale bot back? When was that decided?

**Comment by shaneutt** [comment]
> Why is this stale bot back? When was that decided?

We have the stale bot enabled because an issue being "open" generally means we are either in some state of triage, or the issue has some level of prioritization.

Currently we don't have any prioritization for this issue (we the maintainers have limited capacity, and the project often relies on the community asking for features to also be the ones to help deliver) and it would appear we need people from the community to step forward and champion it if we are going to make this happen.

We would ask that if you intend to bump an issue out of stale/rotten that you do so when you are ready to be assigned the issue and become one of it's champions to move it forward.

@ShadowJonathan (or others here) please let us know: is this an issue that you want to take on? We will be happy to support your efforts: please put it on the agenda for the next SIG Network community meeting so we can discuss it with you!

**Comment by ShadowJonathan** [comment]
I don't know if I have the capacity to work on the issue, but it is a feature that would be helpful to some ideas that came across in the past (that'd be worked on with kubernetes' scale-to-zero), and/or I know that the wider community would have another useful tool in their toolbox that could make kubernetes a useful replacement over already-existing linux workload managers, such as SystemD.

**Comment by sftim** [comment]
If nobody is planning to pick this up, we should put back the stale marker.

**Comment by Zetanova** [comment]
The issue is not the implementation itself.
It requires an extension in the k8s core pod definition and  lifecycle.  

What the community would need is a plan in what steps and how to implement it. 

I researched this issue extensively and got to the conclusion,
First It can be split in 4 separate features:

.) **Extension of the Pod definition for Init-Container**
   A init-container need to be possible to be declared as long-running.
   This would require the init-container to signal the pod that the initialization has been completed,
   but **without** using the return code and **exiting the process**.  
   With this extension it is already possible to make a simple socket-activation init-container

.) **Extension of the Pod definition for Socket Resources.**
     It should be possible to define all types of sockets/pipes inside the Pod definition. (like AF_UNIX, AF_INET ..)
     The k8s-pod can then create them at initializing time (pod-pending) like it is done for volumes/configmap 
     This would mirror the same behavior like systemd does.  [see](https://0pointer.de/blog/projects/socket-activation.html)
     This feature by itself would be already a great addition. container-restarts and container-ipc could already profit from it. 
     Example CSI: kubelet<>sidecars<>csi-driver communiction

.) **Extension of the Pod-Container definition for Startup (container waiting phase)**
   It should be possible to define a container to be scheduled but not to be started until a `run` signal/event.
   The `run` signal/event can contain dynamic generated `ENV` to be used for the startup of the container.  
   This would resolve an other big issue/feature with the [STS ordinal index](https://github.com/kubernetes/kubernetes/issues/40651)
 
.) **Extension of the Pod-Container-Healthcheck** 
   A container should be able to be marked as **static** live and/or ready. 
   This would deactivate the health check until the label is removed. 
   This would enable a service to transfer traffic to the pod for activation

**Comment by sathieu** [comment]
> *Extension of the Pod definition for Init-Container*

This looks like [sidecar containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) which is currently beta.

**Comment by Zetanova** [comment]
@sathieu  Yes, this is exactly the feature, with it socket-activation can be implanted.  

A sidecar container can create a tcp socket and get into ready state on the first connection.
This would postpone the main-container initializing (hopping hat k8s is redownloading the image) 

The only thing that would be missing is to disable/postpone the health check.
Or else the `service` need to have the `publishNotReadyAddresses: true` field set.

**Comment by shaneutt** [comment]
While we appreciate the details and the interest, we still don't seem to have anyone who's stepping forward to be assigned the issue for the moment, so it seems the previous lifecycle was still accurate:

/lifecycle rotten

Please do let us know if you are interested in taking this one on, and we can provide support.

**Comment by wreed4** [comment]
@Zetanova Why is current conversation around this feature talking about "suspending" a pod?? As @thockin said in this issue a year ago, as long as a pod object exists in etcd, you've incurred the cost required to schedule.. at which point "suspending" it does nothing for you.  The value here is specifically in having zero pod objects existing underneath a service and only having a pod being created and scheduled once a web requests comes in.  I apologize if I missed the answer to this question further up in the thread, this is a long one.

**Comment by thockin** [comment]
The idea of a "suspended" pod which has a minimized footprint (perhaps literally 0, perhaps just "less than nominal") is not off the table.  For example, you could have N running pods and M suspended pods - when an up-scale is needed, pick some subset of the suspended pods (1 <= K <= M) to try to "wake up" (in-place resize + state change).  Not all wakeup calls will succeed, but that's OK.  What I just described is NOT particularly related to this issue, though.  It's more of a "make horizontal scaling faster" thing.

True scale-to-zero means scaling in real time to traffic, which means SOMEONE has to hold the traffic, almost certainly BEFORE a destination node has been selected.

**Comment by christf** [comment]
> The idea of a "suspended" pod which has a minimized footprint (perhaps literally 0, perhaps just "less than nominal") is not off the table.  For example, you could have N running pods and M suspended pods - when an up-scale is needed, pick some subset of the suspended pods (1 <= K <= M) to try to "wake up" (in-place resize + state change).  Not all wakeup calls will succeed, but that's OK.  What I just described is NOT particularly related to this issue, though.  It's more of a "make horizontal scaling faster" thing.

Suspending pods means suspending applications which is an entirely different can of worms than scaling. Apps have tasks and event loops that time out, connection pools to refresh etc. To make suspending useful, these resources may have to be refreshed/ reopened 


> 
> True scale-to-zero means scaling in real time to traffic, which means SOMEONE has to hold the traffic, almost certainly BEFORE a destination node has been selected.

What about the ingress controllers. It would seem, they might be in a good position for that.

**Comment by thockin** [comment]
> Suspending pods means suspending applications which is an entirely different can of worms

Sure.  Obviously, if we did something like this, it would be opt-in, and pod owners would know what's going on.

> What about the ingress controllers. It would seem, they might be in a good position for that.

If the traffic you want goes through one, sure.  Pod-to-service traffic doesn't, for example.

**Comment by Zetanova** [comment]
@wreed4 don't know who started with `suspended`. I think there is a miss understanding with the socket-activation function.

I explained it here multiple times and differently. 

The `socket-activation` is a well known feature/pattern of `systemd`. docker itself does support it too. 
It is basically the postponing of the start of a process until a signal of from a new tcp-connection
The process itself takes the created socket `uds` with a well known ENV var. 

After the process/container starts it handles the first connection in the created socket 
and it uses the same socket for all following connections until it decides to exit (scale down to zero)

Of course the POD and IP and resource limits and requests are used.

With the new sidecar init container feature, it should be possible to use socket activation for the main-container:
https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/

The simple benefit is that a idling service does not consume any memory and cpu cycles.
Multiple services can also start in a self defining sequence. 
`systemd` solved with it boot sequence and resource-burst problems.

@sathieu 
socket-activation for sidecars like backup, sshd are still not possible.


"When" I have some time and can update my k8s, I will test it out.

Possible savings:
- dotnet service would save at least 60mb memory and 1-4m idle-cpu.
- PHP container like phpmyadmin 45mb mem, 1m cpu
- kubelet-csr-approver 2x ~20mb mem, 1-2m cpu
- powerdnsadmi  250mb mem, 2m cpu 
- cert-manager-webhook 49mb mem, 4m cpu
- cert-manager maybe 117mb mem, 155m cpu 
- and so on

**Comment by thockin** [comment]
@Zetanova socket-activation works but it depends on the fact that the agent which receives the traffic is able to hand-off to the final process.  When it's all local to a machine, that works pretty well.

What this proposal needs is to flesh out all of the OTHER cases where such hand-off is not obvious.

If a service uses `externalTrafficPolicy: Local` then either: a) the upstream LB can't choose a node to deliver to until the new Pod is scheduled; or b) we sometimes violate the `externalTrafficPolicy`.

If a Pod is talking to a Service in the same cluster, same problem but for `internalTrafficPolicy`.

There are lots of cases to consider.  It may be viable to spec some of them away - e.g. you can't use scale-to-zero for `iTP/eTP: Local`).  As stated elsewhere, this would need a KEP.

**Comment by Zetanova** [comment]
@thockin  socket-activation is `pod internal`. 

It solves the issue of idle-cpu load and real-memory allocation.
Not every process runs smooth like nginx does.  

Many processes using 1-5m idle cpu and >50mb memory to idle around.
Bad once >50m cpu and over >124MB

Sidecar container like sshd or rsync for backup or [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/)  can profit from socket-activation

k8s defined `resources.requests` and under/over commitment is related but a separated issue.

We can make tech demo with the above mentioned sidecar-init-container, this would already be a start.
A KDE for  `resources.requests`, `static health-check status`, `Sockets as a resource` and `Startup signal` can be made severalty.

**Comment by thockin** [comment]
> It solves the issue of idle-cpu load and real-memory allocation.

That's an entirely different problem than "scale-to-zero", and I don't think it needs Kubernetes changes to support it?

If you don't mind pre-scheduling your pods (and committing the `requests`) then what's left seems like it can be worked around within the pod or even within a container:

```
while true {
    accept()
    fork()
    exec(realCommand)
}
```

Obviously there's a little more than that, but I am no longer confident in my understanding of what you think Kubernetes should be providing.

**Comment by Zetanova** [comment]
@thockin 
`systemd` and `docker` both supporting currently the socket-activation pattern

In k8s it should be now possible with ´sidecar-init-container´ to archive near the same.

But there are still view improvements to make regarding health-check, resource.requests and so on.

One extension for the health-check definition,  what I can think of could be:
- Define a container with statically readiness and liveness healthy status and remove the label when the container starts.
- When the defined container exit code is X then don't restart it and set it back to statically readiness and liveness healthy status.

**Comment by thockin** [comment]
I think I see what you are getting at, but I'll be honest - I am doubtful that we want to do this, when in-pod workarounds exist.  Without solving the scheduling/requests part of it, the return-on-complexity doesn't seem justifiable.

As always, prove me wrong - show a demo, write a KEP, convince me I am not seeing the whole picture.
